### PR TITLE
Add FXIOS-3893 FXIOS-3106 [v104] Bookmarks folder decoupling

### PR DESCRIPTION
**Pull request by lmarceau**
# [FXIOS-3893](https://github.com/mozilla-mobile/firefox-ios/issues/10200) [FXIOS-3106](https://github.com/mozilla-mobile/firefox-ios/issues/9034)
Pointing to an epic-branch as this is a WIP that will require both https://github.com/mozilla-mobile/firefox-ios/issues/10233 and https://github.com/mozilla-mobile/firefox-ios/issues/11140 to be brought to `main`.
- Added a protocol `FxBookmarkNode` so we can decouple ourselves from A-S folder structure and make changes locally to how our folder appears to the user.
- Changed folder structure so root now appears to be mobile folder. We add a local desktop folder "Desktop Bookmarks" so user can access their desktop folders. 
- When clicking on that local desktop folder, the following happens: We create folder that appears to be unfiled, toolbar and menu folders. This avoid us from making three different backend calls to get folder names that we already have locally. 
- When we click on unfiled, toolbar and menu folders we fetch the folders content as usual so "faking" those folders locally has no effect. Just avoiding extra calls as seen in https://github.com/mozilla-mobile/firefox-ios/commit/d09700603740931eda63328fab76ea5f47ef5ec5
- Note: Edit action on the toolbar isn't proper at the moment as it will need #10233 and #11140. When we're in the mobile folder it needs to be editable, and when we're in the desktop folder it shouldn't be editable.
- Did some clean up in the Bookmarks panel so we use our latests protocols such as `Loggable`, `ReusableCell`, `Notifiable` as well as `ImageIdentifiers` and `AccessibilityIdentifiers`.
- Also fixed indentation in most places so we're under 120 char and splitted some stuff up with MARK so it's clearer what's happening in there.

### CODE REVIEW

File: Client/Frontend/Library/Bookmarks/BookmarksPanel.swift [hunk 1]
```diff
@@ -151,57 +111,153 @@ class BookmarksPanel: SiteTableViewController, LibraryPanel, CanRemoveQuickActio
...
 
+    // MARK: - Data
+
     override func reloadData() {
         // Can be called while app backgrounded and the db closed, don't try to reload the data source in this case
         if profile.isShutdown { return }
-        profile.places.getBookmarksTree(rootGUID: bookmarkFolderGUID, recursive: false).uponQueue(.main) { result in
 
+        if bookmarkFolderGUID == BookmarkRoots.RootGUID {
+            setupRootFolderData()
```
Comment (yoanarios) [inline Client/Frontend/Library/Bookmarks/BookmarksPanel.swift:121]: 😍 for the separate setup functions way easier to understand the reload function now

File: Client/Frontend/Library/Bookmarks/BookmarksPanel.swift [hunk 2]
```diff
@@ -151,57 +111,153 @@ class BookmarksPanel: SiteTableViewController, LibraryPanel, CanRemoveQuickActio
...
-                self.tableView.reloadData()
+            self.tableView.reloadData()
+
+            self.flashRowIfNeeded()
+        }
+    }
+
+    // MARK: - Actions
+
+    func presentInFolderActions() {
```
Comment (yoanarios) [inline Client/Frontend/Library/Bookmarks/BookmarksPanel.swift:186]: Nice, again easier to read and understand

**Comment by lmarceau** [comment]
Failing tests relates to https://github.com/mozilla-mobile/firefox-ios/issues/11161

File: Client/Frontend/Library/Bookmarks/BookmarksPanel.swift [hunk 3]
```diff
@@ -33,39 +23,52 @@ private class SeparatorTableViewCell: OneLineTableViewCell {
...
+    }
+
+    private enum BookmarksSection: Int, CaseIterable {
         case bookmarks
-        case recent
     }
 
+    // MARK: - Properties
     var libraryPanelDelegate: LibraryPanelDelegate?
+    var notificationCenter = NotificationCenter.default
```
[the remark this answers is not in the document]
Comment (lmarceau) [reply Client/Frontend/Library/Bookmarks/BookmarksPanel.swift:39]: Unless we have a protocol to mock I wouldn't add it as a dependency init, we use default notification center even in tests.

File: Client/Frontend/Library/Bookmarks/BookmarksPanel.swift [hunk 4]
```diff
@@ -151,57 +111,153 @@ class BookmarksPanel: SiteTableViewController, LibraryPanel, CanRemoveQuickActio
...
+
+        } else {
+            setupSubfolderData()
+        }
+    }
+
+    /// "Root" folder is showing mobile bookmarks with a local desktop folder
+    private func setupRootFolderData() {
+        profile.places.getBookmarksTree(rootGUID: BookmarkRoots.MobileFolderGUID,
+                                        recursive: false).uponQueue(.main) { result in
```
[the remark this answers is not in the document]
Comment (lmarceau) [reply Client/Frontend/Library/Bookmarks/BookmarksPanel.swift:134]: Nit about this style, closing bracket placement looks weird. There's an extra space when using the automatic Ctrl+i indentation with Xcode when I set the indentation as suggested. I'm not sure how I feel about it

Screen Shot 2022-06-30 at 10 59 08 AM

**Comment by lmarceau** [comment]
Failing live push tests were solved with https://github.com/mozilla-mobile/firefox-ios/pull/11173. Since this branch is pointing to an epic branch it's not up to date. I'll rebase the epic branch after the merge.

**Pull request merged**